zk_prover/src/merkle_sum_tree/node.rs

```diff
@@ -1,34 +1,55 @@
+use crate::chips::poseidon::poseidon_spec::PoseidonSpec;
+use crate::merkle_sum_tree::utils::big_uint_to_fp;
+use halo2_gadgets::poseidon::primitives::{self as poseidon, ConstantLength};
 use halo2_proofs::halo2curves::bn256::Fr as Fp;
 use num_bigint::BigUint;
 
-use super::{
-    big_uint_to_fp,
-    utils::{poseidon_entry, poseidon_node},
-};
-
 #[derive(Clone, Debug)]
 pub struct Node<const N_ASSETS: usize> {
     pub hash: Fp,
     pub balances: [Fp; N_ASSETS],
 }
 impl<const N_ASSETS: usize> Node<N_ASSETS> {
+    /// Builds a leaf-level node of the MST
+    /// The leaf node hash is equal to `H(username, balance[0], balance[1], ... balance[N_ASSETS - 1])`
+    /// The balances are equal to `balance[0], balance[1], ... balance[N_ASSETS - 1]`
+    pub fn leaf(username: &BigUint, balances: &[BigUint; N_ASSETS]) -> Node<N_ASSETS>
```

Comment:
In the true spirit of DRY this could just call the `Node::leaf_node_from_preimage()` after converting the args :) Same with the middle calling Node::middle_from_preimage inside itself. Won't request changes, we can do this minor enhancement later.

Reply:
Let's DRY! 

Gonna apply the changes now (: